### DISCUSSION TITLE
Multi-threading with pdfplumber

### DISCUSSION THREAD

**Discussion by CheesyZaem** [jsvine/pdfplumber#814] [Q&A]
Hi. I am having some performance issue with extracting texts. I am trying to integrate pdfplumber with tkinter but it seems that Page.extract_text is blocking my main thread from updating the GUI despite running the function in another thread. Below is an example for testing, where the progress bar will stop updating/moving when text is being extracted.

```
import customtkinter
import pdfplumber
from threading import Thread

path = "file.pdf"

app = customtkinter.CTk()
app.geometry("300x300")

def read_pdf():
    with pdfplumber.open(path) as pdf:
        data_pages = pdf.pages

        texts = [page.extract_text() for page in data_pages]
        
    print(texts)
    return texts

frame_1 = customtkinter.CTkFrame(master=app)
frame_1.pack(pady=20, padx=60, fill="both", expand=True)

progressbar_1 = customtkinter.CTkProgressBar(master=frame_1)
progressbar_1.pack(pady=10, padx=10)
progressbar_1.start()

thread = Thread(target=read_pdf, args=())

thread.start()

app.mainloop()
```

It would be much appreciated if there is any way to prevent text extraction from blocking GUI loop.

**Comment by jsvine** [comment]
Ah, very interesting. A few questions:

- Does the blocking still occur if you remove `print(texts)` from the command?
- Does the blocking still occur if you move `path = "file.pdf"` to inside `read_pdf()`?
- Does the blocking still occur if you replace `texts = [page.extract_text() for page in data_pages]` with `time.sleep(10)`?

**Reply by CheesyZaem** [reply]
For your questions, the blocking still occurs for the first two scenarios except the last one.
I am actually not exactly sure if it is pdfplumber completely blocking the main thread because while extraction is in progress, the GUI may sometimes still update by 1 or 2 frames in between.

**Reply by jsvine** [reply]
Thank you for these details, @CheesyZaem. I don't know enough about threading (or `customtkinter`) to diagnose the issue definitively, but perhaps someone else in the community could provide some suggestions.
